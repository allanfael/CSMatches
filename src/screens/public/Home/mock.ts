import { Match } from 'interfaces/matches'

export const matchesMock: Match[] = [
  {
    id: '1',
    time: '2024-01-09T14:00:00Z',
    live: false,
    teamOne: {
      name: 'teamOne',
      image: '',
    },
    teamTwo: {
      name: 'teamTwo',
      image: '',
    },
    league: {
      name: 'leagueTest',
      image: '',
    },
  },
]
