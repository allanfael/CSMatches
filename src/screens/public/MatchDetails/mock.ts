import { MatchDetail } from 'interfaces/match-detail'

export const matchDetailsMock: MatchDetail = {
  teamOne: {
    name: 'Nixuh',
    image:
      'https://cdn.pandascore.co/images/team/image/131795/nixuhlogo_square.png',
  },
  teamTwo: {
    name: 'Goliath',
    image:
      'https://cdn.pandascore.co/images/team/image/126559/goliath_gaming.png',
  },
  time: '2024-01-09T14:00:00Z',
  league: 'PGL Major',
  teamOnePlayers: [
    {
      name: 'Alexey',
      image:
        'https://cdn.pandascore.co/images/player/image/22702/900px_nickel_back_at_dh_dehli_2019.png',
      nickname: 'NickelBack',
    },
    {
      name: 'Vladislav',
      image:
        'https://cdn.pandascore.co/images/player/image/23232/600px_krad_at_champions_cup_finals.png',
      nickname: 'Krad',
    },
  ],
  teamTwoPlayers: [
    {
      name: 'Aleksandr',
      image: 'https://cdn.pandascore.co/images/player/image/24697/p24809.png',
      nickname: 'TRAVIS',
    },
    {
      name: 'Igor',
      image:
        'https://cdn.pandascore.co/images/player/image/30826/600px_forester_at_champions_cup_finals.png',
      nickname: 'Forester',
    },
  ],
}
