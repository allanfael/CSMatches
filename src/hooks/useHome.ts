import { useCallback, useEffect, useRef, useState } from 'react'
import { RouterProps } from '@navigator/ParamsRoute'
import { useNavigation } from '@react-navigation/native'
import { api } from '@services/api/matches'
import { MatchDetail } from 'interfaces/match-detail'
import { Match } from 'interfaces/matches'
import { ROUTERS } from 'utils/routers'

import { matchDetailsMock } from '../screens/public/MatchDetails/mock'

export type UseHomeProps = {
  matches: Match[]
  loading: boolean
  loadingFetchMore: boolean
  refresh: boolean
  onFetchMore(): void
  goToDetails(value: Match): void
  update(): void
}

export const useHome = (): UseHomeProps => {
  const [loading, setLoading] = useState(false)
  const [loadingFetchMore, setLoadingFetchMore] = useState(false)
  const [refresh, setRefresh] = useState(false)
  const [matches, setMatches] = useState<Match[]>([])
  const [page, setPage] = useState(1)
  const isFirstRender = useRef(true)

  const { navigate } = useNavigation<RouterProps>()

  const fetch = useCallback(async () => {
    try {
      setLoading(true)
      const data = await api.matches(1)
      setMatches(data)
    } catch (error) {
    } finally {
      setLoading(false)
      setTimeout(() => {
        isFirstRender.current = false
      }, 100)
    }
  }, [])

  const update = useCallback(async () => {
    try {
      setRefresh(true)
      const data = await api.matches(1)
      setMatches(data)
    } catch (error) {
    } finally {
      setRefresh(false)
    }
  }, [])

  useEffect(() => {
    fetch()
  }, [fetch])

  const onFetchMore = useCallback(async () => {
    if (loadingFetchMore || isFirstRender.current) return

    const nextPage = page + 1

    try {
      setLoadingFetchMore(true)
      const data = await api.matches(nextPage)
      setPage((prev) => prev + 1)
      setMatches((prev) => [...prev, ...data])
    } catch (error) {
    } finally {
      setLoadingFetchMore(false)
    }
  }, [loadingFetchMore, page])

  const goToDetails = (match: Match) => {
    const matchDetails: MatchDetail = {
      time: match.time,
      league: match.league.name,
      teamOne: match.teamOne,
      teamTwo: match.teamTwo,
      teamOnePlayers: matchDetailsMock.teamOnePlayers,
      teamTwoPlayers: matchDetailsMock.teamTwoPlayers,
    }

    navigate(ROUTERS.DETAILS, { match: matchDetails })
  }

  return {
    matches,
    loading,
    onFetchMore,
    loadingFetchMore,
    goToDetails,
    refresh,
    update,
  }
}
