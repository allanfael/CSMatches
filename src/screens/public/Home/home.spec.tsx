import { useHome, UseHomeProps } from '@hooks/useHome'
import { fireEvent, render, screen } from '@testing-library/react-native'

import { matchesMock } from './mock'
import { Home } from '.'

jest.mock('../../../hooks/useHome', () => ({
  useHome: jest.fn(),
}))

describe('<Home />', () => {
  let sut: JSX.Element

  beforeAll(() => {
    sut = <Home />
  })

  it('should render Home correctly', async () => {
    ;(useHome as jest.MockedFunction<typeof useHome>).mockReturnValue({
      loading: false,
      loadingFetchMore: false,
      onFetchMore: jest.fn(),
      goToDetails: jest.fn(),
      matches: matchesMock,
    } as UseHomeProps)

    render(sut)

    expect(screen.toJSON()).toMatchSnapshot()
  })

  it('should render loading component correctly', async () => {
    ;(useHome as jest.MockedFunction<typeof useHome>).mockReturnValue({
      loading: true,
      loadingFetchMore: false,
      onFetchMore: jest.fn(),
      goToDetails: jest.fn(),
      matches: matchesMock,
    } as UseHomeProps)

    const { getByTestId } = render(sut)

    expect(getByTestId('test-loading')).toBeDefined()
  })

  it('should render list correctly', async () => {
    ;(useHome as jest.MockedFunction<typeof useHome>).mockReturnValue({
      loading: false,
      loadingFetchMore: false,
      onFetchMore: jest.fn(),
      goToDetails: jest.fn(),
      matches: matchesMock,
    } as UseHomeProps)

    const { getByText } = render(sut)

    const teamOneName = getByText('teamOne')
    const teamTwoName = getByText('teamTwo')
    const leagueName = getByText('leagueTest')

    expect(teamOneName).toBeTruthy()
    expect(teamTwoName).toBeTruthy()
    expect(leagueName).toBeTruthy()
  })

  it('should render fetchMore component', async () => {
    ;(useHome as jest.MockedFunction<typeof useHome>).mockReturnValue({
      loading: false,
      loadingFetchMore: true,
      onFetchMore: jest.fn(),
      goToDetails: jest.fn(),
      matches: matchesMock,
    } as UseHomeProps)

    const { getByTestId } = render(sut)

    expect(getByTestId('test-loading-fetchMore')).toBeDefined()
  })

  it('should call goToDetails with item', async () => {
    const mockPressedFn = jest.fn()

    ;(useHome as jest.MockedFunction<typeof useHome>).mockReturnValue({
      loading: false,
      loadingFetchMore: false,
      onFetchMore: jest.fn(),
      goToDetails: mockPressedFn,
      matches: matchesMock,
    } as UseHomeProps)

    const { getByText } = render(sut)

    fireEvent.press(getByText('teamOne'))

    expect(mockPressedFn).toHaveBeenCalledWith(matchesMock[0])
  })
})
