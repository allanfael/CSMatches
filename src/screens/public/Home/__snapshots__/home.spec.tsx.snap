// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> should render Home correctly 1`] = `
<View
  style={
    {
      "marginTop": 0,
    }
  }
>
  <View
    style={
      {
        "flex": 1,
        "overflow": "hidden",
      }
    }
  >
    <RCTScrollView
      ItemSeparatorComponent={[Function]}
      ListFooterComponent={[Function]}
      ListHeaderComponent={[Function]}
      applyWindowCorrection={[Function]}
      canChangeSize={true}
      contentContainerStyle={
        {
          "backgroundColor": undefined,
          "minHeight": 1,
          "minWidth": 1,
          "paddingLeft": 0,
          "paddingRight": 0,
        }
      }
      contentHeight={50}
      contentWidth={400}
      data={
        [
          {
            "id": "1",
            "league": {
              "image": "",
              "name": "leagueTest",
            },
            "live": false,
            "teamOne": {
              "image": "",
              "name": "teamOne",
            },
            "teamTwo": {
              "image": "",
              "name": "teamTwo",
            },
            "time": "2024-01-09T14:00:00Z",
          },
        ]
      }
      dataProvider={
        DataProvider {
          "_data": [
            {
              "id": "1",
              "league": {
                "image": "",
                "name": "leagueTest",
              },
              "live": false,
              "teamOne": {
                "image": "",
                "name": "teamOne",
              },
              "teamTwo": {
                "image": "",
                "name": "teamTwo",
              },
              "time": "2024-01-09T14:00:00Z",
            },
          ],
          "_firstIndexToProcess": 0,
          "_hasStableIds": true,
          "_requiresDataChangeHandling": false,
          "_size": 1,
          "getStableId": [Function],
          "rowHasChanged": [Function],
        }
      }
      disableRecycling={false}
      estimatedItemSize={50}
      extendedState={{}}
      externalScrollView={[Function]}
      finalRenderAheadOffset={250}
      forceNonDeterministicRendering={true}
      horizontal={false}
      initialOffset={0}
      initialRenderIndex={0}
      isHorizontal={false}
      keyExtractor={[Function]}
      layoutProvider={
        GridLayoutProviderWithProps {
          "_acceptableRelayoutDelta": 1,
          "_getHeightOrWidth": [Function],
          "_getLayoutTypeForIndex": [Function],
          "_getSpan": [Function],
          "_hasExpired": false,
          "_isHorizontal": false,
          "_lastLayoutManager": GridLayoutManager {
            "_acceptableRelayoutDelta": 1,
            "_getSpan": [Function],
            "_isGridHorizontal": false,
            "_isHorizontal": false,
            "_layoutProvider": [Circular],
            "_layouts": [
              {
                "height": 50,
                "type": 0,
                "width": 400,
                "x": 0,
                "y": 0,
              },
            ],
            "_maxSpan": 1,
            "_renderWindowSize": {
              "height": 1000,
              "width": 400,
            },
            "_totalHeight": 50,
            "_totalWidth": 400,
            "_window": {
              "height": 1000,
              "width": 400,
            },
          },
          "_maxSpan": 1,
          "_renderWindowSize": {
            "height": 1000,
            "width": 400,
          },
          "_setLayoutForType": [Function],
          "_tempDim": {
            "height": 0,
            "width": 0,
          },
          "averageWindow": AverageWindow {
            "currentAverage": 50,
            "currentCount": 1,
            "inputValues": [
              50,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
            ],
            "nextIndex": 1,
          },
          "defaultEstimatedItemSize": 100,
          "layoutObject": {
            "size": undefined,
            "span": undefined,
          },
          "props": {
            "ItemSeparatorComponent": [Function],
            "ListFooterComponent": [Function],
            "ListFooterComponentStyle": undefined,
            "ListHeaderComponent": [Function],
            "ListHeaderComponentStyle": undefined,
            "contentContainerStyle": undefined,
            "data": [
              {
                "id": "1",
                "league": {
                  "image": "",
                  "name": "leagueTest",
                },
                "live": false,
                "teamOne": {
                  "image": "",
                  "name": "teamOne",
                },
                "teamTwo": {
                  "image": "",
                  "name": "teamTwo",
                },
                "time": "2024-01-09T14:00:00Z",
              },
            ],
            "estimatedItemSize": 50,
            "estimatedListSize": {
              "height": 1000,
              "width": 400,
            },
            "horizontal": false,
            "keyExtractor": [Function],
            "numColumns": 1,
            "onEndReached": [MockFunction] {
              "calls": [
                [],
              ],
              "results": [
                {
                  "type": "return",
                  "value": undefined,
                },
              ],
            },
            "onEndReachedThreshold": 0.2,
            "renderItem": [Function],
            "testID": "test-flashlist",
          },
          "renderWindowInsets": {
            "height": 0,
            "width": 0,
          },
          "shouldRefreshWithAnchoring": true,
        }
      }
      layoutSize={
        {
          "height": 1000,
          "width": 400,
        }
      }
      maxRenderAhead={750}
      numColumns={1}
      onEndReached={[Function]}
      onEndReachedThreshold={0}
      onEndReachedThresholdRelative={0.2}
      onItemLayout={[Function]}
      onLayout={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onSizeChanged={[Function]}
      onVisibleIndicesChanged={[Function]}
      removeClippedSubviews={false}
      renderAheadOffset={0}
      renderAheadStep={250}
      renderContentContainer={[Function]}
      renderItem={[Function]}
      renderItemContainer={[Function]}
      rowRenderer={[Function]}
      scrollEventThrottle={16}
      scrollThrottle={16}
      scrollViewProps={
        {
          "contentContainerStyle": {
            "backgroundColor": undefined,
            "minHeight": 1,
            "minWidth": 1,
            "paddingLeft": 0,
            "paddingRight": 0,
          },
          "onLayout": [Function],
          "onScrollBeginDrag": [Function],
          "refreshControl": undefined,
          "style": {
            "minHeight": 1,
            "minWidth": 1,
          },
        }
      }
      style={
        {
          "minHeight": 1,
          "minWidth": 1,
        }
      }
      suppressBoundedSizeException={true}
      testID="test-flashlist"
      windowCorrectionConfig={
        {
          "applyToInitialOffset": true,
          "applyToItemScroll": true,
          "value": {
            "endCorrection": 0,
            "startCorrection": 0,
            "windowShift": -1,
          },
        }
      }
    >
      <View>
        <View
          style={
            {
              "flexDirection": "column",
            }
          }
        >
          <View
            style={
              {
                "paddingLeft": 0,
                "paddingTop": 0,
              }
            }
          />
          <View
            style={
              [
                undefined,
                undefined,
              ]
            }
          >
            <Text
              style={
                {
                  "color": "#fff",
                  "fontFamily": "Roboto_500Medium",
                  "fontSize": 32,
                  "lineHeight": 38,
                }
              }
            >
              Partidas
            </Text>
          </View>
          <AutoLayoutView
            enableInstrumentation={false}
            horizontal={false}
            onBlankAreaEvent={[Function]}
            onLayout={[Function]}
            renderAheadOffset={0}
            scrollOffset={0}
            style={
              {
                "height": 50,
                "width": 400,
              }
            }
            windowSize={1000}
          >
            <CellContainer
              index={0}
              onLayout={[Function]}
              style={
                {
                  "alignItems": "stretch",
                  "flexDirection": "column",
                  "left": 0,
                  "position": "absolute",
                  "top": 0,
                  "width": 400,
                }
              }
            >
              <View
                style={
                  {
                    "flexDirection": "column",
                  }
                }
              >
                <View
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                >
                  <View
                    style={
                      {
                        "backgroundColor": "#FAFAFA33",
                      }
                    }
                  >
                    <Text
                      style={
                        {
                          "color": "#fff",
                          "fontFamily": "Roboto_400Regular",
                          "fontSize": 8,
                          "lineHeight": 12,
                        }
                      }
                    >
                      Today at 11:00 AM
                    </Text>
                  </View>
                  <View>
                    <View>
                      <View>
                        <ViewManagerAdapter_ExpoImage
                          accessibilityLabel="image"
                          backgroundColor={4291085508}
                          contentFit="contain"
                          contentPosition={
                            {
                              "left": "50%",
                              "top": "50%",
                            }
                          }
                          onError={[Function]}
                          onLoad={[Function]}
                          onLoadStart={[Function]}
                          onProgress={[Function]}
                          placeholder={[]}
                          source={
                            [
                              {
                                "uri": "",
                              },
                            ]
                          }
                          style={
                            {
                              "backgroundColor": "#C4C4C4",
                            }
                          }
                          transition={null}
                        />
                      </View>
                      <Text
                        style={
                          {
                            "color": "#fff",
                            "fontFamily": "Roboto_400Regular",
                            "fontSize": 10,
                            "lineHeight": 16,
                          }
                        }
                      >
                        teamOne
                      </Text>
                    </View>
                    <Text
                      style={
                        {
                          "color": "#C4C4C4",
                          "fontFamily": "Roboto_400Regular",
                          "fontSize": 12,
                          "lineHeight": 18,
                        }
                      }
                    >
                      vs
                    </Text>
                    <View>
                      <View>
                        <ViewManagerAdapter_ExpoImage
                          accessibilityLabel="image"
                          backgroundColor={4291085508}
                          contentFit="contain"
                          contentPosition={
                            {
                              "left": "50%",
                              "top": "50%",
                            }
                          }
                          onError={[Function]}
                          onLoad={[Function]}
                          onLoadStart={[Function]}
                          onProgress={[Function]}
                          placeholder={[]}
                          source={
                            [
                              {
                                "uri": "",
                              },
                            ]
                          }
                          style={
                            {
                              "backgroundColor": "#C4C4C4",
                            }
                          }
                          transition={null}
                        />
                      </View>
                      <Text
                        style={
                          {
                            "color": "#fff",
                            "fontFamily": "Roboto_400Regular",
                            "fontSize": 10,
                            "lineHeight": 16,
                          }
                        }
                      >
                        teamTwo
                      </Text>
                    </View>
                  </View>
                  <View />
                  <View>
                    <View>
                      <ViewManagerAdapter_ExpoImage
                        accessibilityLabel="image"
                        backgroundColor={4291085508}
                        contentFit="contain"
                        contentPosition={
                          {
                            "left": "50%",
                            "top": "50%",
                          }
                        }
                        onError={[Function]}
                        onLoad={[Function]}
                        onLoadStart={[Function]}
                        onProgress={[Function]}
                        placeholder={[]}
                        source={
                          [
                            {
                              "uri": "",
                            },
                          ]
                        }
                        style={
                          {
                            "backgroundColor": "#C4C4C4",
                          }
                        }
                        transition={null}
                      />
                    </View>
                    <Text
                      style={
                        {
                          "color": "#fff",
                          "fontFamily": "Roboto_400Regular",
                          "fontSize": 8,
                          "lineHeight": 12,
                        }
                      }
                    >
                      leagueTest
                    </Text>
                  </View>
                </View>
              </View>
            </CellContainer>
          </AutoLayoutView>
          <CellContainer
            index={-1}
            style={
              [
                undefined,
                undefined,
              ]
            }
          />
          <View
            style={
              {
                "paddingBottom": 0,
                "paddingRight": 0,
              }
            }
          />
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
